drivers/i2c/i2c_omap.c

```diff
@@ -0,0 +1,748 @@
+ */
+
+static int i2c_omap_recover_bus(const struct device *dev)
+{
+	const struct i2c_omap_cfg *cfg = DEV_CFG(dev);
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+
+	struct i2c_bitbang bitbang_omap;
+	struct i2c_bitbang_io bitbang_omap_io = {
+		.get_sda = i2c_omap_get_sda,
+		.set_scl = i2c_omap_set_scl,
+		.set_sda = i2c_omap_set_sda,
+	};
+	int error = 0;
+
+	i2c_base_addr->SYSTEST |= I2C_OMAP_SYSTEST_ST_EN | (3 << I2C_OMAP_SYSTEST_TMODE_SHIFT) |
+				  I2C_OMAP_SYSTEST_SCL_O | I2C_OMAP_SYSTEST_SDA_O;
+	i2c_bitbang_init(&bitbang_omap, &bitbang_omap_io, (void *)cfg);
+	error = i2c_bitbang_recover_bus(&bitbang_omap);
+	if (error != 0) {
+		LOG_ERR("failed to recover bus (err %d)", error);
+		goto restore;
+	}
+
+restore:
+	i2c_base_addr->SYSTEST &= ~(I2C_OMAP_SYSTEST_ST_EN | I2C_OMAP_SYSTEST_TMODE_MASK |
+				    I2C_OMAP_SYSTEST_SCL_O | I2C_OMAP_SYSTEST_SDA_O);
+	i2c_omap_reset(dev);
+	return error;
+}
+#endif /* CONFIG_I2C_OMAP_BUS_RECOVERY */
+
+/**
+ * @brief Wait for the bus to become free (no longer busy).
+ *
+ * This function waits for the bus to become free by continuously checking the
+ * status register of the OMAP I2C controller. If the bus remains busy for a
+ * certain timeout period, the function will return attempts to recover the bus by calling
+ * i2c_omap_recover_bus().
+ *
+ * @param dev The I2C device structure.
+ * @return 0 if the bus becomes free, or a negative error code if the bus cannot
+ * be recovered.
+ */
+static int i2c_omap_wait_for_bb(const struct device *dev)
+{
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+	uint32_t timeout = k_uptime_get_32() + I2C_OMAP_TIMEOUT;
+
+	while (i2c_base_addr->STAT & I2C_OMAP_STAT_BB) {
+		if (k_uptime_get_32() > timeout) {
+			LOG_ERR("Bus busy timeout");
+#ifdef CONFIG_I2C_OMAP_BUS_RECOVERY
+			return i2c_omap_recover_bus(dev);
+#else
+			return -ETIMEDOUT;
+#endif /* CONFIG_I2C_OMAP_BUS_RECOVERY */
+		}
+		k_busy_wait(100);
+	}
+	return 0;
+}
+
+/**
+ * @brief Performs data transfer for the OMAP I2C driver.
+ *
+ * This function is responsible for handling the data transfer logic for the OMAP I2C driver.
+ * It reads the status register and performs the necessary actions based on the status flags.
+ * It handles both receive and transmit logic, and also handles error conditions such as NACK,
+ * arbitration lost, receive overrun, and transmit underflow.
+ *
+ * @param dev Pointer to the device structure.
+ *
+ * @return Returns 0 on success, or a negative error code on failure.
+ */
+static int i2c_omap_transfer_message_ll(const struct device *dev)
+{
+	struct i2c_omap_data *data = DEV_DATA(dev);
+	i2c_omap_regs_t *i2c_base_addr = DEV_I2C_BASE(dev);
+	uint16_t stat = i2c_base_addr->STAT;
+
+	if (data->receiver) {
+		stat &= ~(I2C_OMAP_STAT_XDR | I2C_OMAP_STAT_XRDY);
+	} else {
+		stat &= ~(I2C_OMAP_STAT_RDR | I2C_OMAP_STAT_RRDY);
+	}
+	if (stat & I2C_OMAP_STAT_NACK) {
+		data->cmd_err = -I2C_OMAP_STAT_NACK;
```

Comment:
Here you are assigning `-I2C_OMAP_STAT_NACK`, but below you check for `data->cmd_err & I2C_OMAP_STAT_NACK`. Should this line be `data->cmd_err |= I2C_OMAP_STAT_NACK;`?

Same below for `I2C_OMAP_STAT_AL`, but the issue would be you have defined your return code `RETRY` to be the same bit value.

Better question I now have is why do we need `cmd_err` at all? Why not just return the error directly from this function instead of storing is and fetching it out later in the calling function?